backend/src/apiserver/template/v2_template.go

```diff
@@ -306,3 +311,55 @@ func IsPlatformSpecWithKubernetesConfig(template []byte) bool {
 	_, ok := platformSpec.Platforms["kubernetes"]
 	return ok
 }
+
+func (t *V2Spec) validatePipelineJobInputs(job *pipelinespec.PipelineJob) error {
+	// The pipeline requires no input
+	if t.spec.GetRoot().GetInputDefinitions() == nil {
+		return nil
+	}
+
+	runtimeConfig := job.GetRuntimeConfig()
+	if runtimeConfig == nil {
+		return fmt.Errorf("pipeline requiring input has empty runtime config")
+	}
+
+	for name, param := range t.spec.GetRoot().GetInputDefinitions().GetParameters() {
+		if input, ok := runtimeConfig.GetParameterValues()[name]; !ok {
+			// If the parameter is optional, or there is a default value, it's ok to not have a user input
+			if !param.GetIsOptional() && param.GetDefaultValue() == nil {
+				return fmt.Errorf("parameter %s is not optional, yet has neither default value nor user provided value", name)
+			}
+		} else {
+			// Verify the parameter type is correct
+			switch param.GetParameterType() {
+			case pipelinespec.ParameterType_PARAMETER_TYPE_ENUM_UNSPECIFIED:
+				return fmt.Errorf("input parameter %s has unspecified type", name)
+			case pipelinespec.ParameterType_NUMBER_DOUBLE, pipelinespec.ParameterType_NUMBER_INTEGER:
+				if _, ok := input.GetKind().(*structpb.Value_NumberValue); !ok {
+					return fmt.Errorf("input parameter %s has type double or integer, but the input parameter is not of Value_NumberValue type", name)
```

Comment:
nit: would probably avoid using `Value_NumberValue` in the user facing error message.

Reply:
I understand that `Value_NumberValue` is more of some internal info, but it feels helpful that that users know what the required type is. Certainly it would be much better if the types of parameter and inputs are represented in the same way.

Reply:
Sorry, I didn't make it very clear here. On the left it should say 
<pre>
input parameter %s <b>requires</b> type double or integer
</pre>
This is the type the pipeline calls for. This type maps to a `structpb.Value` "subtype", which is the user provided value, such as double/integer mapping to `Value_NumberValue`.